connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```
@@ -16,56 +16,325 @@
+     * store is provided, a secondary write is made to that store if the write to the connector-specific store
+     * succeeds. Errors with this secondary write are not reflected in the returned {@link Future} or the passed-in
+     * {@link Callback}; they are only logged as a warning to users.
+     *
+     * <p>If not configured to use a connector-specific offset store, the returned {@link Future} corresponds to a
+     * write to the worker-global offset store, and the passed-in {@link Callback} is invoked once that write completes.
+
+     * @param values map from key to value
+     * @param callback callback to invoke on completion of the primary write
+     * @return void future for the primary write
+     */
     @Override
     public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callback) {
-        // TODO
-        return workerStore.set(values, callback);
+        final OffsetBackingStore primaryStore;
+        final OffsetBackingStore secondaryStore;
+        if (connectorStore.isPresent()) {
+            primaryStore = connectorStore.get();
+            secondaryStore = workerStore.orElse(null);
+        } else if (workerStore.isPresent()) {
+            primaryStore = workerStore.get();
+            secondaryStore = null;
+        } else {
+            // Should never happen since we check for this case in the constructor, but just in case, this should
+            // be more informative than the NPE that would otherwise be thrown
+            throw new IllegalStateException("At least one non-null offset store must be provided");
+        }
+
+        return primaryStore.set(values, (primaryWriteError, ignored) -> {
+            if (secondaryStore != null) {
+                if (primaryWriteError != null) {
+                    log.trace("Skipping offsets write to secondary store because primary write has failed", primaryWriteError);
```

Why did we log `trace` to this primary store error, but `warn` to the secondary store error in L290?

---

The failure to write to the primary store will probably be logged elsewhere at a higher level, since we report it to the callback. We swallow failures to write to the secondary store, so this is the only chance we get to log those.

---

I see. Thanks.